src/test/java/software/amazon/nio/spi/s3/FakeAsyncS3ClientBuilder.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package software.amazon.nio.spi.s3;
+
+import java.net.URI;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.S3CrtAsyncClientBuilder;
+import software.amazon.awssdk.services.s3.crt.S3CrtHttpConfiguration;
+import software.amazon.awssdk.services.s3.crt.S3CrtRetryConfiguration;
+
+/**
+ * This class fakes a S3CrtAsyncClientBuilder for testability purposes. It
```

Comment:
I think you should use a Mockito Spy for this purpose (https://www.baeldung.com/mockito-spy)

Reply:
I have tried to do it, but after a while and for the reason below I parked it. My points are:

1. in general, I prefer using code I can fully control and if possible, not use mocks but the real code, so that I am as closest as possible to the real production code;
2. with the Fake class approach, I am actually using a real BUILDER with all the side effects it may bring
3. As far as I understood spying (although I admit I am not an expert), you can mimic method calls, but you can not inject you own custom logic

Reply:
A Spy should use a real class (like a Builder). It can record, intercept and modify calls to the real classes methods but perhaps isn't the best way to inject custom logic. What is the custom logic you need?

Reply:
It's in the functionality of the Fake class: I want to be able to make sure I am passing the right configuration to the real S3Client. As far as I could understand, you can not easily access the config you use in the builder from an instance of a S3Client. I could not even figure out where some parameters where stored, and in some cases I could find them in private fields. A simple use case I wasn't able to figure out how to achieve with a spy:

```
S3ClientStore cs = new S3ClientStore();
cs.generateAsyncClient(...);
assertTrue(URI.create("https://endpoint1.io"), ... client's endpoint... );
```